numpy/lib/function_base.py

```diff
@@ -1616,27 +1619,47 @@ def trim_zeros(filt, trim='fb'):
     >>> np.trim_zeros(a, 'b')
     array([0, 0, 0, ..., 0, 2, 1])
 
+    >>> b = np.array(['', '', '' ,'a', 'b', 'c', '', 'b', 'a', ''])
+    >>> np.trim_zeros(b)
+    array(['a', 'b', 'c', '', 'b', 'a'], dtype='<U1')
+
     The input data type is preserved, list/tuple in means list/tuple out.
 
     >>> np.trim_zeros([0, 1, 2, 0])
     [1, 2]
 
     """
-    first = 0
-    trim = trim.upper()
-    if 'F' in trim:
-        for i in filt:
-            if i != 0.:
-                break
-            else:
-                first = first + 1
-    last = len(filt)
-    if 'B' in trim:
-        for i in filt[::-1]:
-            if i != 0.:
-                break
-            else:
-                last = last - 1
+    try:
+        arr = np.asanyarray(filt, dtype=bool)
+
+    # str/bytes and structured arrays cannot be directly converted into bool arrays
+    except (TypeError, ValueError):
+        arr_any = np.asanyarray(filt)
+        _arr = arr_any.view(bool)
+        _arr.shape = arr_any.shape + (arr_any.dtype.itemsize,)
+        arr = _arr.any(axis=-1)
```

Comment:
I don't think this branch makes sense - I think you want either:
```
    arr = asanyarray(filt) == 0
```
or
```
    arr = asanyarray(filt).astype(bool, copy=False)
```

Reply:
I actually hadn't even considered that; implemented in https://github.com/numpy/numpy/pull/16911/commits/bb21711f862ad5d1212af727440b617d38f8c73b.